File: .NET/Microsoft.Recognizers.Text.Sequence/Extractors/BaseURLExtractor.cs
using System.Collections.Generic;
using System.Collections.Immutable;
using System.Linq;
using System.Text.RegularExpressions;

using Microsoft.Recognizers.Definitions;
using Microsoft.Recognizers.Text.Matcher;

namespace Microsoft.Recognizers.Text.Sequence
{
    public class BaseURLExtractor : BaseSequenceExtractor
    {
        public BaseURLExtractor()
        {
            var regexes = new Dictionary<Regex, string>
            {
                {
                    new Regex(BaseURL.IpUrlRegex, RegexOptions.Compiled),
                    Constants.URL_REGEX
                },
                {
                    new Regex(BaseURL.UrlRegex, RegexOptions.Compiled),
                    Constants.URL_REGEX
                },
                {
                    new Regex(BaseURL.UrlRegex2, RegexOptions.Compiled),
                    Constants.URL_REGEX
                },
            };

            Regexes = regexes.ToImmutableDictionary();
            AmbiguousTimeTerm = new Regex(BaseURL.AmbiguousTimeTerm, RegexOptions.Compiled);

            TldMatcher = new StringMatcher();
            TldMatcher.Init(BaseURL.TldList);
        }

        internal override ImmutableDictionary<Regex, string> Regexes { get; }

        protected sealed override string ExtractType { get; } = Constants.SYS_URL;

        private StringMatcher TldMatcher { get; }

        private Regex AmbiguousTimeTerm { get; }

        public override bool IsValidMatch(Match match)
        {
            var isValidTld = false;
            var isIPUrl = match.Groups["IPurl"].Success;

            if (!isIPUrl)
            {
                var tldString = match.Groups["Tld"].Value;
                var tldMatches = TldMatcher.Find(tldString);

                if (tldMatches.Any(o => o.Start == 0 && o.End == tldString.Length))
                {
                    isValidTld = true;
                }
            }

            // For cases like "7.am" or "8.pm" which are more likely time terms.
            if (AmbiguousTimeTerm.IsMatch(match.Value))
            {
                return false;
            }

            return isValidTld || isIPUrl;
        }
    }
}

File: Java/libraries/recognizers-text-sequence/src/main/java/com/microsoft/recognizers/text/sequence/extractors/BaseURLExtractor.java
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

package com.microsoft.recognizers.text.sequence.extractors;

import com.microsoft.recognizers.text.matcher.MatchResult;
import com.microsoft.recognizers.text.matcher.StringMatcher;
import com.microsoft.recognizers.text.sequence.Constants;
import com.microsoft.recognizers.text.sequence.config.URLConfiguration;
import com.microsoft.recognizers.text.sequence.resources.BaseURL;
import com.microsoft.recognizers.text.utilities.Match;

import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.regex.Pattern;
import java.util.stream.Collectors;
import java.util.stream.StreamSupport;

public class BaseURLExtractor extends BaseSequenceExtractor {
    private final URLConfiguration config;
    private StringMatcher tldMatcher;
    private Pattern ambiguousTimeTerm;

    protected final String extractType = Constants.SYS_URL;

    protected Map<Pattern, String> getRegexes() {
        return regexes;
    }

    protected String getExtractType() {
        return extractType;
    }

    public BaseURLExtractor(URLConfiguration config) {
        this.config = config;
        Map<Pattern, String> regexes = new HashMap<Pattern, String>() {
            {
                put(config.getUrlRegex(), Constants.URL_REGEX);
                put(config.getIpUrlRegex(), Constants.URL_REGEX);
                put(Pattern.compile(BaseURL.UrlRegex2), Constants.URL_REGEX);
            }
        };

        super.regexes = regexes;
        this.ambiguousTimeTerm = Pattern.compile(BaseURL.AmbiguousTimeTerm);

        this.tldMatcher = new StringMatcher();
        this.tldMatcher.init(BaseURL.TldList);
    }

    @Override
    public Boolean isValidMatch(Match match) {
        // For cases like "7.am" or "8.pm" which are more likely time terms.
        return !this.ambiguousTimeTerm.matcher(match.value).find();
    }
}
